File: ios-native/NeobankNative/Sources/Features/Transfers/TransferRepository.swift
import Foundation

struct TransferRepository: Sendable {
    let client: APIClient

    /// A 422 here means "declined" — the gateway returns a full transfer
    /// body (with `status`/`failure_reason` set), not an error envelope. The
    /// generic error path doesn't have this body, so decode it from
    /// `APIError.responseData` instead of surfacing a bare failure.
    func createTransfer(
        idempotencyKey: String,
        amount: String,
        recipientPhone: String?,
        recipientEmail: String?,
        recipientUserId: String?,
        currency: String?,
        memo: String?
    ) async throws -> Transfer {
        do {
            return try await client.send(
                "/v1/transfers",
                method: .post,
                body: [
                    "amount": amount,
                    "recipient_phone": recipientPhone?.isEmpty == false ? recipientPhone : nil,
                    "recipient_email": recipientEmail?.isEmpty == false ? recipientEmail : nil,
                    "recipient_user_id": recipientUserId?.isEmpty == false ? recipientUserId : nil,
                    "currency": currency?.isEmpty == false ? currency : nil,
                    "memo": memo?.isEmpty == false ? memo : nil,
                ],
                idempotencyKey: idempotencyKey
            )
        } catch let error as APIError where error.statusCode == 422 {
            if let data = error.responseData, let transfer = try? APIClient.jsonDecoder.decode(Transfer.self, from: data) {
                return transfer
            }
            throw error
        }
    }
}

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/transfers/TransferRepository.kt
package dev.horobets.neobanknative.features.transfers

import dev.horobets.neobanknative.core.networking.ApiClient
import dev.horobets.neobanknative.core.networking.ApiError
import dev.horobets.neobanknative.core.networking.HttpMethod
import dev.horobets.neobanknative.core.networking.jsonBodyOf
import kotlinx.serialization.decodeFromString

class TransferRepository(private val client: ApiClient) {
    /**
     * A 422 here means "declined" — the gateway returns a full transfer
     * body (with `status`/`failure_reason` set), not an error envelope. The
     * generic error path doesn't have this body, so decode it from
     * [ApiError.responseBody] instead of surfacing a bare failure.
     */
    suspend fun createTransfer(
        idempotencyKey: String,
        amount: String,
        recipientPhone: String?,
        recipientEmail: String?,
        recipientUserId: String?,
        currency: String?,
        memo: String?,
    ): Transfer = try {
        client.send(
            "/v1/transfers",
            method = HttpMethod.POST,
            body = jsonBodyOf(
                "amount" to amount,
                "recipient_phone" to recipientPhone?.ifEmpty { null },
                "recipient_email" to recipientEmail?.ifEmpty { null },
                "recipient_user_id" to recipientUserId?.ifEmpty { null },
                "currency" to currency?.ifEmpty { null },
                "memo" to memo?.ifEmpty { null },
            ),
            idempotencyKey = idempotencyKey,
        )
    } catch (e: ApiError) {
        val body = e.responseBody
        if (e.statusCode == 422 && body != null) {
            try {
                ApiClient.json.decodeFromString<Transfer>(body)
            } catch (decodeError: Exception) {
                throw e
            }
        } else {
            throw e
        }
    }
}
